Adding User Datasource to Morpheus subprovider

RDavisHPE opened the pull request:
Adding the User datasource to the Morpheus subprovider so that data about users can be queried in Terraform configurations using the `data "hpe_morpheus_user"` block.
- Implemented datasource interface in datasource.go
- Added user datasource to datasources.go
- Defined access objects in Morpheus API spec and updated hpe-morpheus-go-sdk and hpe-morpheus-provider codespec with new OpenAPI version respectively.
- Ran go get github.com/HewlettPackard/hpe-morpheus-go-sdk@main and go mod tidy for SDK bump
- Ran make docs for documentation

RDavisHPE committed:
Adding User Datasource to Morpheus subprovider

Adding the User datasource to the Morpheus subprovider so that data about
users can be queried in Terraform configurations using the `data "hpe_morpheus_user"` block.
- Implemented datasource interface in datasource.go
- Added user datasource to datasources.go
- Moved consts to seperate file
- Defined access objects in Morpheus API spec and updated hpe-morpheus-go-sdk and hpe-morpheus-provider-codespec
  with new OpenAPI version respectively.
- Ran go get github.com/HewlettPackard/hpe-morpheus-go-sdk@main and go mod tidy for SDK bump
- Ran make docs for documentation

RDavisHPE commented on internal/subproviders/morpheus/datasources/user/datasource.go:
```diff
@@ -0,0 +1,540 @@
+		buildAppTemplates := func() types.Set {
+			appTemplatesValues := []attr.Value{}
+			for _, in := range access.GetAppTemplates() {
+				appTemplatesValues = append(appTemplatesValues,
```
Was curious if I could use convert.ToSetType for these but looking at examples from instance it would seem this involves referencing SDK struct names explicitly (which we would like to avoid in case they change):
```go
tags, d := convert.ToSetType(
	ctx,
	resp.GetInstance().Tags,
	func(
		in sdk.AddInstance200ResponseAllOfOneOfInstanceTagsInner,
	) TagsValue {
		return TagsValue{
			Name:  convert.StrToType(in.Name),
			Value: convert.StrToType(in.Value),
			state: attr.ValueStateKnown,
		}
	},
)
```

matthew-c-hpe commented:
Could you separate out the sdk version bump into a separate PR?

matthew-c-hpe commented on internal/subproviders/morpheus/datasources/user/schema_gen.go:
```diff
@@ -0,0 +1,8249 @@
+			"linux_key_pair_id": schema.Int64Attribute{
+				Computed: true,
+			},
+			"linux_password": schema.StringAttribute{
```
I don't think there's much point in having linux or windows passwords available in the data source given that the API hides them as `"************"` if they're set.

RDavisHPE replied:
Might still be useful in case someone just wants to check whether those values are set (null check), but I don’t feel strongly about it.

matthew-c-hpe commented on internal/subproviders/morpheus/datasources/user/schema_gen.go:
```diff
@@ -0,0 +1,8249 @@
+	Id                   types.Int64         `tfsdk:"id"`
+	IncludeAccess        types.Bool          `tfsdk:"include_access"`
+	IsUsing2fa           types.Bool          `tfsdk:"is_using2fa"`
+	LastLoginDate        types.String        `tfsdk:"last_login_date"`
```
In a Terraform and Morpheus context, I'm not sure how useful it would be to have the date and time-related properties `LastLoginDate`, `DateCreated`, `LastUpdated` available in the data source. I think we could leave them out.

RDavisHPE replied:
Initially I was sort of thinking it could be useful for aggregating users created before/after a certain time - but just remembered the datasource works pretty explicitly with just a single user - pretty happy removing these.

matthew-c-hpe commented on internal/subproviders/morpheus/datasources/user/schema_gen.go:
```diff
@@ -0,0 +1,8249 @@
+
+var _ basetypes.ObjectValuable = AccessValue{}
+
+type AccessValue struct {
```
In the new provider we're trying to adopt the common naming convention of Morpheus resources as per the UI and Documentation, this means that the schema will have to be adjusted if we will include the access attribute in the data source. See the Role resource implementation for guidance

RDavisHPE replied:
https://github.com/hpe-cds/hpe-morpheus-provider-code-spec/pull/134

matthew-c-hpe commented on internal/subproviders/morpheus/datasources/user/schema_gen.go:
```diff
@@ -0,0 +1,8249 @@
+var _ basetypes.ObjectValuable = AccessValue{}
+
+type AccessValue struct {
+	AppTemplates     basetypes.SetValue `tfsdk:"app_templates"`
```
AppTemplates should be renamed to Blueprints

matthew-c-hpe commented on internal/subproviders/morpheus/datasources/user/schema_gen.go:
```diff
@@ -0,0 +1,8249 @@
+	Personas         basetypes.SetValue `tfsdk:"personas"`
+	ReportTypes      basetypes.SetValue `tfsdk:"report_types"`
+	Sites            basetypes.SetValue `tfsdk:"sites"`
+	TaskSets         basetypes.SetValue `tfsdk:"task_sets"`
```
TaskSets should be renamed to Workflows

matthew-c-hpe commented on internal/subproviders/morpheus/datasources/user/schema_gen.go:
```diff
@@ -0,0 +1,8249 @@
+	TaskSets         basetypes.SetValue `tfsdk:"task_sets"`
+	Tasks            basetypes.SetValue `tfsdk:"tasks"`
+	VdiPools         basetypes.SetValue `tfsdk:"vdi_pools"`
+	Zones            basetypes.SetValue `tfsdk:"zones"`
```
Zones should be renamed to Clouds

matthew-c-hpe commented on internal/subproviders/morpheus/datasources/user/schema_gen.go:
```diff
@@ -0,0 +1,8249 @@
+	InstanceTypes    basetypes.SetValue `tfsdk:"instance_types"`
+	Personas         basetypes.SetValue `tfsdk:"personas"`
+	ReportTypes      basetypes.SetValue `tfsdk:"report_types"`
+	Sites            basetypes.SetValue `tfsdk:"sites"`
```
Sites should be renamed to Groups.

matthew-c-hpe commented on internal/subproviders/morpheus/datasources/user/datasource.go:
```diff
@@ -0,0 +1,540 @@
+	}
+
+	// Handle Access complex object
+	if access, accessOk := user.GetAccessOk(); accessOk {
```
Break the building up of the access object out into a separate function

RDavisHPE replied:
Seperated - with the caveat that we are now using a SDK struct name in the function signature so if that gets renamed it will be a breaking change.

matthew-c-hpe commented on internal/subproviders/morpheus/datasources/user/schema_gen.go:
```diff
@@ -0,0 +1,8249 @@
+	LinuxPassword        types.String        `tfsdk:"linux_password"`
+	LinuxUsername        types.String        `tfsdk:"linux_username"`
+	LoginAttempts        types.Int64         `tfsdk:"login_attempts"`
+	LoginCount           types.Int64         `tfsdk:"login_count"`
```
I don't think having `LoginAttempts` or  `LoginCount` available in the data source is that useful, I think we could leave them out.

matthew-c-hpe commented on internal/subproviders/morpheus/datasources/user/schema_gen.go:
```diff
@@ -0,0 +1,8249 @@
+	Access               AccessValue         `tfsdk:"access"`
+	Account              AccountValue        `tfsdk:"account"`
+	AccountExpired       types.Bool          `tfsdk:"account_expired"`
+	AccountId            types.Int64         `tfsdk:"account_id"`
```
I think this should be renamed to `TenantId` in line with UI/documentation naming conventions.

RDavisHPE commented:
> Could you separate out the sdk version bump into a separate PR?

👍 
https://github.com/HPE/terraform-provider-hpe/pull/332

matthew-c-hpe commented on internal/subproviders/morpheus/datasources/user/datasource.go:
```diff
@@ -0,0 +1,540 @@
+	getUserRequest := apiClient.UsersAPI.GetUser(ctx, id)
+
+	// Add includeAccess parameter if specified
+	if !data.IncludeAccess.IsNull() && data.IncludeAccess.ValueBool() {
```
If you end up leaving in the access field, I think we should simplify the data source schema by omitting `include_access` from it and instead always setting `IncludeAccess` in the request to true here.

That way it will be more consistent with the other data source schemas that we have - so far the only attributes we've allowed the user to set have been those which filter what Morpheus resource the data source returns (`id` and resource `name`).

RDavisHPE replied:
Seems fair - my main reason for including it was that the access blob is huge (kilobytes of data) so the user might not want to download/manage all of that.

RDavisHPE replied:
Defaulting includeAccess query parameter to be true :+1:

RDavisHPE committed:
User Datasource Access Naming Schema Update and fixing include_access to true

RDavisHPE committed:
User Datasource Access Naming Schema Update and fixing include_access to true

Regenerated schema using new naming convention and omitting include_access.
Also now omitting linux_password, windows_password, login_count, login_attempts,
last_updated, last_login_date, date_created.
Removed time to string conversion function as it is no longer needed.

- Regenerated schema_gen.go
- Updated datasoure.go to use new naming and set includeAccess query parameter to true
- Ran make docs to update documentation

RDavisHPE committed:
Moving access value construction to its own function
Also updated naming for access fields in implementation
Changed implementation to reflect populateGetRoleAsStatePermissions()

matthew-c-hpe commented on internal/subproviders/morpheus/datasources/user/datasource.go:
```diff
@@ -0,0 +1,546 @@
+			if len(blueprintsValues) == 0 {
+				return types.SetNull(BlueprintsValue{}.Type(ctx))
+			}
+			set, _ := types.SetValue(BlueprintsValue{}.Type(ctx), blueprintsValues)
```
With the function based implemention you've done here for building these up, we lose out on diagnostics information if the nested access types fail to correctly construct a set. I think this could be made more resilient by exiting early and returning those diagnostics if it fails to construct the nested types.

I'd recommend taking a look at Role's `populateGetRoleAsStatePermissions()` and changing these to a similar pattern so that we can propagate the diagnostics of any failed set conversion to the caller of `getAccessAsState`.

RDavisHPE replied:
Updated - much cleaner now (if a little longer with the added checks)

matthew-c-hpe reviewed (approved):
LGTM, thanks
